include/epanet2_2.h

```diff
@@ -940,6 +933,24 @@ typedef struct Project *EN_Project;
   int DLLEXPORT EN_setdemandmodel(EN_Project ph, int type, double pmin,
                 double preq, double pexp);
 
+  /**
+  @brief Adds a new category of demand data to a junction node.
+  */
+  int DLLEXPORT EN_adddemand(EN_Project p, int nodeIndex, double baseDemand,
+                char *pattern, const char *name, int *key);
+
+  /**
+  @brief Removes a demand category from a junction node.
+  */
+  int DLLEXPORT EN_deletedemand(EN_Project p, int nodeIndex, int demandIndex);
+
+  /**
+  @brief Gets the index of the demand category at a junction node with a specific
+         name or key.
+  */
+  int DLLEXPORT EN_getdemandindex(EN_Project p, int nodeIndex, const char *name,
```

Comment:
Right now the generic list has no random access mechanism -- using random keys is just another way to "stamp" a demand category similar to giving it a name. In general (no pun intended) list containers don't provide random access -- vectors do. If you want to turn the list into a vector then you have to redefine what a node is and this will break the use of the node struct in all list clients. Hence my decision to keep a list node hidden.